KubeArmor/core/kubeUpdate.go

```diff
@@ -509,6 +509,66 @@ func (dm *KubeArmorDaemon) UpdateEndPointWithPod(action string, pod tp.K8sPod) {
 	}
 }
 
+func getTopLevelOwner(obj metav1.ObjectMeta, namespace string, objkind string) (string, string, string, error) {
+	ownerRef := kl.GetControllingPodOwner(obj.OwnerReferences)
+	if ownerRef == nil {
+		return obj.Name, objkind, namespace, nil
+	}
+
+	switch ownerRef.Kind {
+	case "Pod":
+		pod, err := K8s.K8sClient.CoreV1().Pods(namespace).Get(context.Background(), ownerRef.Name, metav1.GetOptions{})
+		if err != nil {
+			return "", "", "", err
+		}
+		if len(pod.OwnerReferences) > 0 {
+			return getTopLevelOwner(pod.ObjectMeta, namespace, "Pod")
+		}
+	case "Deployment":
+		deployment, err := K8s.K8sClient.AppsV1().Deployments(namespace).Get(context.Background(), ownerRef.Name, metav1.GetOptions{})
+		if err != nil {
+			return "", "", "", err
+		}
+		if len(deployment.OwnerReferences) > 0 {
+			return getTopLevelOwner(deployment.ObjectMeta, namespace, "Deployment")
+		}
+		return deployment.Name, "Deployment", deployment.Namespace, nil
+	case "ReplicaSet":
+		replicaset, err := K8s.K8sClient.AppsV1().ReplicaSets(namespace).Get(context.Background(), ownerRef.Name, metav1.GetOptions{})
+		if err != nil {
+			return "", "", "", err
+		}
+		if len(replicaset.OwnerReferences) > 0 {
+			return getTopLevelOwner(replicaset.ObjectMeta, namespace, "ReplicaSet")
+		}
+		return replicaset.Name, "ReplicaSet", replicaset.Namespace, nil
```

Comment:
why deployment case is not handled here?

Reply:
that is handled the function is recusively being called if there is a delpoyment owner it will be handled by case "deployment"

Reply:
ohh yes, my bad, got confused with `objkind`.